java-frontend/src/main/java/org/sonar/java/resolve/Resolve.java

```diff
@@ -424,61 +424,73 @@ private Resolution findMethodInStaticImport(Env env, String name, List<JavaType>
+      return findConstructor(env, site, argTypes, typeParams);
+    } else if ("super".equals(name)) {
+      if (superclass == null) {
+        return unresolved();
+      }
+      return findConstructor(env, superclass, argTypes, typeParams);
+    }
+
+    return findMethodByStrictThenLooseInvocation(env, callSite, site, name, argTypes, typeParams);
   }
 
-  private Resolution findConstructor(Env env, JavaType site, List<JavaType> argTypes, List<JavaType> typeParams, boolean autoboxing) {
+  private Resolution findConstructor(Env env, JavaType site, List<JavaType> argTypes, List<JavaType> typeParams) {
     List<JavaType> newArgTypes = argTypes;
     JavaSymbol owner = site.symbol.owner();
     if (!owner.isPackageSymbol() && !site.symbol.isStatic()) {
       // JLS8 - 8.8.1 & 8.8.9 : constructors of inner class have an implicit first arg of its directly enclosing class type
       newArgTypes = ImmutableList.<JavaType>builder().add(owner.enclosingClass().type).addAll(argTypes).build();
     }
-    return findMethod(env, site, site, CONSTRUCTOR_NAME, newArgTypes, typeParams, autoboxing);
+    return findMethodByStrictThenLooseInvocation(env, site, site, CONSTRUCTOR_NAME, newArgTypes, typeParams);
   }
 
-  private Resolution findMethod(Env env, JavaType callSite, JavaType site, String name, List<JavaType> argTypes, List<JavaType> typeParams, boolean autoboxing) {
+  private Resolution findMethodByStrictThenLooseInvocation(Env env, JavaType callSite, JavaType site, String name, List<JavaType> argTypes, List<JavaType> typeParams) {
+    // JLS8 - §5.3 searching by strict invocation, then loose invocation
+    Resolution bestSoFar = findMethod(env, callSite, site, name, argTypes, typeParams, false);
+    if (bestSoFar.symbol.kind >= JavaSymbol.ERRONEOUS && !argTypes.isEmpty()) {
+      // retry with loose invocation
+      bestSoFar = findMethod(env, callSite, site, name, argTypes, typeParams, true);
+    }
+    return bestSoFar;
+  }
+
+  private Resolution findMethod(Env env, JavaType callSite, JavaType site, String name, List<JavaType> argTypes, List<JavaType> typeParams, boolean looseInvocation) {
     JavaType superclass = site.getSymbol().getSuperclass();
     Resolution bestSoFar = unresolved();
-    // handle constructors
-    if ("this".equals(name)) {
-      return findConstructor(env, site, argTypes, typeParams, autoboxing);
-    } else if ("super".equals(name)) {
-      if (superclass == null) {
-        return bestSoFar;
-      }
-      return findConstructor(env, superclass, argTypes, typeParams, autoboxing);
-    }
-    bestSoFar = lookupInScope(env, callSite, site, name, argTypes, typeParams, autoboxing, site.getSymbol().members(), bestSoFar);
+
+    bestSoFar = lookupInScope(env, callSite, site, name, argTypes, typeParams, looseInvocation, site.getSymbol().members(), bestSoFar);
+    // FIXME SONARJAVA-2096: interrupt exploration if the most specific method has already been found by strict invocation context
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)